File: packages/invoice/src/calculate.ts
import Big from "big.js";

import type {
  InvoiceLineInput,
  InvoiceDiscountInput,
  InvoiceLineResult,
  InvoiceCalculationResult,
} from "./types";

export function calculateInvoice(
  items: InvoiceLineInput[],
  globalDiscount?: InvoiceDiscountInput | null,
): InvoiceCalculationResult {
  const ZERO = new Big(0);
  const HUNDRED = new Big(100);
  let subtotalGross = ZERO;
  // 1. Zwischensumme (nach Artikelrabatten)
  const lineGrossValues: Big[] = [];

  items.forEach((item) => {
    // product * quantity
    let lineGross = new Big(item.priceGross).times(item.quantity);

    // minus discount
    if (item.discount) {
      if (item.discount.type === "fixed") {
        lineGross = lineGross.minus(item.discount.value);
      } else {
        lineGross = lineGross.minus(
          lineGross.times(item.discount.value).div(HUNDRED),
        );
      }
    }

    // check value is minimum 0
    if (lineGross.lt(ZERO)) lineGross = ZERO;
    lineGrossValues.push(lineGross);
    
    subtotalGross = subtotalGross.plus(lineGross);
  });

  // 2. Global Discount
  let finalTotalGross = subtotalGross;
  let discountAmount = ZERO;

  if (globalDiscount) {
    if (globalDiscount.type === "fixed") {
      discountAmount = new Big(globalDiscount.value);
      finalTotalGross = finalTotalGross.minus(discountAmount);
    } else {
      discountAmount = finalTotalGross.times(globalDiscount.value).div(HUNDRED);
      finalTotalGross = finalTotalGross.minus(discountAmount);
    }
  }

  if (finalTotalGross.lt(ZERO)) finalTotalGross = ZERO;
  if (discountAmount.gt(subtotalGross)) discountAmount = subtotalGross;

  // 3. Steuer rückrechnen (gruppiert nach Rate)
  const discountRatio = subtotalGross.gt(ZERO)
    ? finalTotalGross.div(subtotalGross)
    : new Big(1);

  const taxByRate = new Map<string, { gross: Big }>();
  const computedItems: InvoiceLineResult[] = [];

  items.forEach((item, i) => {
    const lineGross = lineGrossValues[i]!;
    const lineGrossAfterGlobal = lineGross.times(discountRatio);
    const ratePct = new Big(item.taxRate).toFixed(2);
    const rate = new Big(item.taxRate).div(100);
    const rowNetRounded = new Big(
      lineGrossAfterGlobal.div(new Big(1).plus(rate)).toFixed(8),
    );

    const prev = taxByRate.get(ratePct) ?? { gross: ZERO };
    taxByRate.set(ratePct, {
      gross: prev.gross.plus(lineGrossAfterGlobal),
    });

    const priceGrossUnit = new Big(item.priceGross);
    computedItems.push({
      productId: item.productId,
      productName: item.productName,
      quantity: item.quantity,
      priceGrossUnit: priceGrossUnit.toFixed(4),
      taxRateSnapshot: ratePct,
      priceNetUnitPrecise: priceGrossUnit.div(new Big(1).plus(rate)).toFixed(8),
      rowTotalGross: lineGrossAfterGlobal.toFixed(2),
      rowTotalNetPrecise: rowNetRounded.toFixed(8),
      discountType: item.discount?.type ?? null,
      discountValue: item.discount
        ? new Big(item.discount.value).toFixed(4)
        : null,
      costCenter: item.costCenter ?? null,
    });
  });

  // 4. Gruppen-Brutto runden + Cent-Korrektur
  const grossRounded = new Big(finalTotalGross.toFixed(2));
  const sorted = Array.from(taxByRate.entries()).sort(([a], [b]) =>
    new Big(a).cmp(new Big(b)),
  );

  const groupGrossRounded = sorted.map(([, v]) =>
    new Big(v.gross.toFixed(2)),
  );
  const groupGrossSum = groupGrossRounded.reduce((s, g) => s.plus(g), ZERO);
  const centDiff = grossRounded.minus(groupGrossSum);

  if (!centDiff.eq(ZERO)) {
    // Cent-Differenz auf die größte Gruppe anwenden
    let largestIdx = 0;
    for (let j = 1; j < groupGrossRounded.length; j++) {
      if (groupGrossRounded[j]!.gt(groupGrossRounded[largestIdx]!)) {
        largestIdx = j;
      }
    }
    groupGrossRounded[largestIdx] = groupGrossRounded[largestIdx]!.plus(centDiff);
  }

  // 5. Steuer pro Gruppe: net + tax = gruppenBrutto (exakt)
  const taxBreakdown = sorted.map(([rate], i) => {
    const gGross = groupGrossRounded[i]!;
    const rateDecimal = new Big(rate).div(HUNDRED);
    const net = new Big(gGross.div(new Big(1).plus(rateDecimal)).toFixed(2));
    const tax = gGross.minus(net);
    return { rate, net: net.toFixed(2), tax: tax.toFixed(2) };
  });

  // 6. Totale aus Breakdown ableiten
  const totalNet = taxBreakdown.reduce(
    (sum, entry) => sum.plus(entry.net),
    ZERO,
  );
  const totalTax = taxBreakdown.reduce(
    (sum, entry) => sum.plus(entry.tax),
    ZERO,
  );

  return {
    subtotal: subtotalGross.toFixed(2),
    discountTotal: discountAmount.toFixed(2),
    gross: grossRounded.toFixed(2),
    net: totalNet.toFixed(2),
    tax: totalTax.toFixed(2),
    taxBreakdown,
    items: computedItems,
    count: items.reduce((sum, item) => sum + item.quantity, 0),
    discountType: globalDiscount?.type ?? null,
    discountValue: globalDiscount
      ? new Big(globalDiscount.value).toFixed(4)
      : null,
  };
}

File: apps/pulpo-app/invoice/calculate.go
package invoice

import (
	"sort"

	"github.com/shopspring/decimal"
)

var (
	zero    = decimal.Zero
	hundred = decimal.NewFromInt(100)
)

// CalculateInvoice computes totals, tax breakdown and per-item results.
// This is a 1:1 port of packages/invoice/src/calculate.ts.
func CalculateInvoice(items []LineInput, globalDiscount *DiscountInput) CalculationResult {
	subtotalGross := zero
	lineGrossValues := make([]decimal.Decimal, len(items))

	// 1. Per-item gross (after item discounts)
	for i, item := range items {
		lineGross := item.PriceGross.Mul(decimal.NewFromInt(int64(item.Quantity)))

		if item.Discount != nil {
			if item.Discount.Type == "fixed" {
				lineGross = lineGross.Sub(item.Discount.Value)
			} else {
				lineGross = lineGross.Sub(lineGross.Mul(item.Discount.Value).Div(hundred))
			}
		}

		if lineGross.LessThan(zero) {
			lineGross = zero
		}

		lineGrossValues[i] = lineGross
		subtotalGross = subtotalGross.Add(lineGross)
	}

	// 2. Global discount
	finalTotalGross := subtotalGross
	discountAmount := zero

	if globalDiscount != nil {
		if globalDiscount.Type == "fixed" {
			discountAmount = globalDiscount.Value
			finalTotalGross = finalTotalGross.Sub(discountAmount)
		} else {
			discountAmount = finalTotalGross.Mul(globalDiscount.Value).Div(hundred)
			finalTotalGross = finalTotalGross.Sub(discountAmount)
		}
	}

	if finalTotalGross.LessThan(zero) {
		finalTotalGross = zero
	}
	if discountAmount.GreaterThan(subtotalGross) {
		discountAmount = subtotalGross
	}

	// 3. Tax back-calculation grouped by rate
	discountRatio := decimal.NewFromInt(1)
	if subtotalGross.GreaterThan(zero) {
		discountRatio = finalTotalGross.Div(subtotalGross)
	}

	type taxGroup struct {
		gross decimal.Decimal
	}
	taxByRate := make(map[string]*taxGroup)
	computedItems := make([]LineResult, len(items))

	for i, item := range items {
		lineGross := lineGrossValues[i]
		lineGrossAfterGlobal := lineGross.Mul(discountRatio)
		ratePct := item.TaxRate.StringFixed(2)

		if _, ok := taxByRate[ratePct]; !ok {
			taxByRate[ratePct] = &taxGroup{gross: zero}
		}
		taxByRate[ratePct].gross = taxByRate[ratePct].gross.Add(lineGrossAfterGlobal)

		lr := LineResult{
			ProductID:       item.ProductID,
			ProductName:     item.ProductName,
			Quantity:        item.Quantity,
			PriceGrossUnit:  item.PriceGross.StringFixed(4),
			TaxRateSnapshot: ratePct,
			RowTotalGross:   lineGrossAfterGlobal.StringFixed(2),
		}

		if item.Discount != nil {
			dt := item.Discount.Type
			dv := item.Discount.Value.StringFixed(4)
			lr.DiscountType = &dt
			lr.DiscountValue = &dv
		}

		if item.CostCenter != "" {
			cc := item.CostCenter
			lr.CostCenter = &cc
		}

		computedItems[i] = lr
	}

	// 4. Round group gross + cent correction
	grossRounded := finalTotalGross.Round(2)

	// Sort rates ascending
	rates := make([]string, 0, len(taxByRate))
	for rate := range taxByRate {
		rates = append(rates, rate)
	}
	sort.Slice(rates, func(i, j int) bool {
		ri, _ := decimal.NewFromString(rates[i])
		rj, _ := decimal.NewFromString(rates[j])
		return ri.LessThan(rj)
	})

	groupGrossRounded := make([]decimal.Decimal, len(rates))
	groupGrossSum := zero
	for i, rate := range rates {
		groupGrossRounded[i] = taxByRate[rate].gross.Round(2)
		groupGrossSum = groupGrossSum.Add(groupGrossRounded[i])
	}

	centDiff := grossRounded.Sub(groupGrossSum)
	if !centDiff.IsZero() {
		largestIdx := 0
		for j := 1; j < len(groupGrossRounded); j++ {
			if groupGrossRounded[j].GreaterThan(groupGrossRounded[largestIdx]) {
				largestIdx = j
			}
		}
		groupGrossRounded[largestIdx] = groupGrossRounded[largestIdx].Add(centDiff)
	}

	// 5. Tax per group: net + tax = groupGross
	taxBreakdown := make([]TaxBreakdownEntry, len(rates))
	for i, rate := range rates {
		gGross := groupGrossRounded[i]
		rateDecimal, _ := decimal.NewFromString(rate)
		rateDiv := rateDecimal.Div(hundred)
		net := gGross.Div(decimal.NewFromInt(1).Add(rateDiv)).Round(2)
		tax := gGross.Sub(net)

		taxBreakdown[i] = TaxBreakdownEntry{
			Rate: rate,
			Net:  net.StringFixed(2),
			Tax:  tax.StringFixed(2),
		}
	}

	// 6. Totals from breakdown
	totalNet := zero
	totalTax := zero
	for _, entry := range taxBreakdown {
		n, _ := decimal.NewFromString(entry.Net)
		t, _ := decimal.NewFromString(entry.Tax)
		totalNet = totalNet.Add(n)
		totalTax = totalTax.Add(t)
	}

	// Count
	count := 0
	for _, item := range items {
		count += item.Quantity
	}

	result := CalculationResult{
		Subtotal:      subtotalGross.StringFixed(2),
		DiscountTotal: discountAmount.StringFixed(2),
		Gross:         grossRounded.StringFixed(2),
		Net:           totalNet.StringFixed(2),
		Tax:           totalTax.StringFixed(2),
		TaxBreakdown:  taxBreakdown,
		Items:         computedItems,
		Count:         count,
	}

	if globalDiscount != nil {
		dt := globalDiscount.Type
		dv := globalDiscount.Value.StringFixed(4)
		result.DiscountType = &dt
		result.DiscountValue = &dv
	}

	return result
}
